crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,733 @@
+use async_trait::async_trait;
+use c8y_api::smartrest::message::collect_smartrest_messages;
+use c8y_api::smartrest::message::get_smartrest_template_id;
+use c8y_api::smartrest::smartrest_deserializer::SmartRestFirmwareRequest;
+use c8y_api::smartrest::smartrest_deserializer::SmartRestRequestGeneric;
+use c8y_api::smartrest::smartrest_serializer::TryIntoOperationStatusMessage;
+use c8y_api::smartrest::topic::C8yTopic;
+use c8y_http_proxy::handle::C8YHttpProxy;
+use mqtt_channel::Topic;
+use nanoid::nanoid;
+use sha256::digest;
+use sha256::try_digest;
+use std::collections::HashMap;
+use std::fs;
+use std::fs::File;
+use std::os::unix::fs as unix_fs;
+use std::path::Path;
+use std::path::PathBuf;
+use tedge_actors::fan_in_message_type;
+use tedge_actors::Actor;
+use tedge_actors::ChannelError;
+use tedge_actors::DynSender;
+use tedge_actors::LoggingReceiver;
+use tedge_actors::MessageBox;
+use tedge_actors::ReceiveMessages;
+use tedge_actors::RuntimeError;
+use tedge_actors::RuntimeRequest;
+use tedge_actors::WrappedInput;
+use tedge_api::health::health_status_up_message;
+use tedge_api::OperationStatus;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_timer_ext::SetTimeout;
+use tedge_timer_ext::Timeout;
+use tracing::error;
+use tracing::info;
+use tracing::warn;
+
+use crate::config::FirmwareManagerConfig;
+use crate::error::FirmwareManagementError;
+use crate::message::get_child_id_from_child_topic;
+use crate::message::DownloadFirmwareStatusMessage;
+use crate::message::FirmwareOperationRequest;
+use crate::message::FirmwareOperationResponse;
+use crate::operation::ActiveOperationState;
+use crate::operation::FirmwareOperationEntry;
+use crate::operation::OperationKey;
+
+pub type OperationTimer = SetTimeout<OperationKey>;
+pub type OperationTimeout = Timeout<OperationKey>;
+
+fan_in_message_type!(FirmwareInput[MqttMessage, OperationTimeout] : Debug);
+fan_in_message_type!(FirmwareOutput[MqttMessage, OperationTimer] : Debug);
+
+pub struct FirmwareManagerActor {
+    config: FirmwareManagerConfig,
+    active_child_ops: HashMap<OperationKey, ActiveOperationState>,
+}
+
+#[async_trait]
+impl Actor for FirmwareManagerActor {
+    type MessageBox = FirmwareManagerMessageBox;
+
+    fn name(&self) -> &str {
+        "FirmwareManager"
+    }
+
+    async fn run(mut self, mut message_box: Self::MessageBox) -> Result<(), RuntimeError> {
+        self.resend_operations_to_child_device(&mut message_box)
+            .await?;
+        // TODO: Do we really need to send 500 from each actor?
+        self.get_pending_operations_from_cloud(&mut message_box)
+            .await?;
```

Comment:
Good point!

It would be good to introduce an actor that manages the handshake with Cumulicity, on start up but also when the bridge is down then up. In order to reduce the number of actors this can be the same actor that would notify C8Y the health status of thin-edge.